### PR TITLE
Final draft for initial version

### PR DESCRIPTION
**Pull request by mscappini**
Once this is merged, I can use it in the web projects.

### CODE REVIEW

**Comment by aisthesis** [comment]
Could you provide a bit more context for the usage in the README (or GitHub wiki, which we get for free)? I'm guessing that you actually will never (?) need to install this directly on a server but only declare (?) it as a dependency in your code, then require it. Could you provide a bit more explanation since it differs from the normal pattern of how to use a service?

**Comment by aisthesis** [comment]
That's the only thing I see for this iteration. I still want to look into how we might use `npm config` rather than an environment variable. The main downside of an environment variable is that if for some reason `npm` should decide to use exactly that variable, we'd have to make some fundamental changes. I'm less worried about `NODE_ENV` because node actually expects it the way we're using it, so they're unlikely to make changes that would break us. Don't worry about this for now. I'm just calling it out because I'd still like to look into that alternative myself, but I don't want to block you in moving forward with `ui` and `restapi`.

**Comment by aisthesis** [comment]
A 'Testing' section should also be included in the documentation, with the 3 sections shown in 'Template'. Does the code pass the conformity tests of `jshint` and `jscs`? There should also be a 'Connects to' section to show that you don't have to worry about that here.

**Comment by mscappini** [comment]
To be clear: I should include the line from the package.json on how to reference this as a dependency, and then the name of the expected environment variable?

File: lib/cfgloader.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        path = require('path'),
+        nconf = require('nconf');
+    
+    var configPath = process.env.NPM_CONF_PATH;
+    if (!configPath) {
+        Object.defineProperty(nconf, 'warning', { value: new Array(1) });
+        nconf.warning[0] = "NPM_CONF_PATH not set";
+    }
+    
+    /* The order in which you attach these configuration sources determines their priority in the hierarchy. */
```
Comment (aisthesis) [inline lib/cfgloader.js]: Shouldn't these comments be deleted?

File: lib/cfgloader.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        }
+        configFile += ".json";
+        
+        // Service config values
+        var envConfig = path.join(configPath, configFile);
+        if (fs.existsSync(envConfig)) {
+            nconf.file(envConfig);
+        }
+        
+        //// Default config values
```
Comment (aisthesis) [inline lib/cfgloader.js]: Same as above: presumably delete.

**Comment by aisthesis** [comment]
It's up to you exactly how, but the following need to be clarified:
1. You _don't_ install this package the way you do our other ones with `npm install`, etc.
2. Where exactly the code snippet you provide is supposed to go.
3. What the code snippet is doing.

**Comment by mscappini** [comment]
"The main downside of an environment variable is that if for some reason npm should decide to use exactly that variable, we'd have to make some fundamental changes." That's why I want to call it something like OB_CONF_PATH.

**Comment by mscappini** [comment]
Please have another look.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	var host = config.get('web:host');
+	var port = config.get('web:port');
+
+This code will read an environment variable named `NPM_CONF_PATH` for a directory to locate a configuration file for the service qualified by name in the `cfgloader` constructor. If environment variable `NPM_CONF_PATH` has a value of `/etc/`, and the first argument provided to `cfgloader`'s constructor is `restapi`, `cfgloader` will look for a configuration file at `/etc/confg-restapi.json` and load it for values to override the `restapi`'s service configuration defaults provided as the second argument to the `cfgloader`'s constructor.
+
+Testing
+---
+
+### Code conformity
+    $ jshint lib test
```
Comment (aisthesis) [inline README.md]: Omit test since there's no test directory.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+		}
+	};
+
+	var cfgloader = require('@optbot/cfgloader');
+	var config = cfgloader(serviceName, defaults);  
+
+	var host = config.get('web:host');
+	var port = config.get('web:port');
+
+This code will read an environment variable named `NPM_CONF_PATH` for a directory to locate a configuration file for the service qualified by name in the `cfgloader` constructor. If environment variable `NPM_CONF_PATH` has a value of `/etc/`, and the first argument provided to `cfgloader`'s constructor is `restapi`, `cfgloader` will look for a configuration file at `/etc/confg-restapi.json` and load it for values to override the `restapi`'s service configuration defaults provided as the second argument to the `cfgloader`'s constructor.
```
Comment (aisthesis) [inline README.md]: I'd make the client responsible for isolating the configurations within a directory. That way you don't have to prefix this in code every time. Following the example, `NPM_CONF_PATH` might be `/etc/npmconfig` and the config file for `restapi` would be `/etc/npmconfig/restapi.json`.

**Comment by aisthesis** [comment]
Looks good.

**Pull request merged**